docs/internal/compiler-steps.md

```diff
@@ -0,0 +1,119 @@
+# Compiler Steps
+
+## Overview
+
+This is a preliminary study on each compiler step to produce executable objects from Verona code.
+The initial objective of the Verona compiler is to generate a single executable from a number of files, directories and existing foreing objects.
+Loadable code and separate compilation processes are not taken into account in this document but are under discussion and may be introduced in the future.
+
+### Verona Sources
+
+The Verona compiler treats all files inside the same directory as a single module.
+Point the compiler to the directory and it will compile all (Verona) files together and include other modules (via `use` or `type`) in the same way.
+
+The end result is a single meta-module containing all modules: the main directory and all included modules from any module.
+This final module must contain a `Main` class with a `main(env : Env) : I32` method on any of the sub-modules, that will be the entry point for the program.
+Having zero or more than one `Main` classes is a compiler error.
+
+### Foreign Code
+
+Verona can also interoperate with foreign code, via sandbox regions and remote calls into the foreign object (archive, shared object, etc).
+
+The main goal of the interoperability layer is to use _functionality in foreign modules_, not to use _foreign code_ inside Verona.
+
+Check the `interop.md` document for more details.
+
+Interoperability code can be found at `src/interop`.
+
+## Parser
+
+The parser is responsible for:
+* Understanding how to navigate directory structures to include all reachable code
+* Recognising foreign calls and trigger parsing of foreing code
+* Check syntax and semantics correctness of all sources
+* Parse the code as one meta-module and emit A-normal form AST for the next stage
+
+The final product of the parser is a fully checked concrete AST that the MLIR layer can consume as is (with potential additional helper structures) and generate plain MLIR without additional checks.
+
+The stages (and status) of the parser are:
+* (mostly done) Lex/Parse (+ syntax check)
+* (mostly done) Resolve symbols
+* (in progress) Type inference + Generics + Reification
+* (in progress) Reachability analysis & code elision
+* (mostly done) Emit A-normal form + Symbol table
+
+The parser code can be found at `src/parser`.
+
+## MLIR
+
+The MLIR stage takes in the AST and symbol table above and lowers to the following MLIR dialects:
+* Standard: Overall structure (functions, modules, etc) + existing operations
+* LLVM: Structures, memory management, remaining operations
+* Region (new): A dialect that tags objects into regions + creation & destruction semantics.
+
+With the (high-level) IR above, the following MLIR passes are executed:
+* Region and alias analysis
+* Heap to stack transformation
+* Foreign code RPC generation
+
+The final lowering to LLVM IR is done via:
+* Partial lowering from the Region dialect to Std/LLVM dialects
+* Final lowering to LLVM dialect and LLVM IR
```

Comment:
Can we list the high-level IR features that we need to elide here?  For example, lowering an abstract method-invoke to a load from the type descriptor and an indirect call.

Reply:
That's done at the parser level. The AST will have two types of calls:
* A direct call, with a fully qualified name and arguments
* An indirect call, with a descriptor and an offset (vtable) and the arguments

Both types of calls are guaranteed to exist and argument/return to type check.

Reply:
For the record, the `Region` dialect is something I just came up with now, as an example of a Verona-related-but-not-specific dialect. Since we can have as many dialects as we want, I think it makes more sense to have a `Region` dialect for the region analysis, and other dialects for different purposes than a `Verona` dialect for all those things.

Future dialects, once we move some of the logic into MLIR, would be for example `Type Algebra` (for type inference), `Call` dialect, for late dynamic vs static choice, along with selector colouring, etc. Those are interesting research projects on themselves, but shouldn't get in the way of reaching a functional compiler first.